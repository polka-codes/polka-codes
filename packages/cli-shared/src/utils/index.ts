export * from './checkRipgrep'
export * from './eventHandler'
export * from './listFiles'
export * from './readMultiline'
export * from './searchFiles'
