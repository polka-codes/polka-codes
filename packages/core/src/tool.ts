export type ToolParameter = {
  name: string
  description: string
  required: boolean
  usageValue?: string
  allowMultiple?: boolean
  children?: ToolParameter[]
}

export type ToolParameterValue = string | { [key: string]: ToolParameterValue } | ToolParameterValue[]

export type ToolExample = {
  description: string
  parameters: { name: string; value: ToolParameterValue }[]
}

export enum PermissionLevel {
  // This tool is completely safe
  None = 0,
  // This tool can read files
  Read = 1,
  // This tool can write files
  Write = 2,
  // This tool can perform arbitrary action
  Arbitrary = 3,
}

export type ToolInfo = {
  name: string
  description: string
  parameters: ToolParameter[]
  examples?: ToolExample[]
  permissionLevel: PermissionLevel
}

export type FullToolInfo = ToolInfo & {
  handler: ToolHandler<ToolInfo, any>
  isAvailable: (provider: any) => boolean
}

export enum ToolResponseType {
  Reply = 'Reply',
  Exit = 'Exit',
  Invalid = 'Invalid',
  Error = 'Error',
  Interrupted = 'Interrupted',
  HandOver = 'HandOver',
  Delegate = 'Delegate',
  Pause = 'Pause',
}

// Reply to the tool use
export type ToolResponseReply = {
  type: ToolResponseType.Reply
  message: string
}

// Should end the message thread
// e.g. task completed successfully
export type ToolResponseExit = {
  type: ToolResponseType.Exit
  message: string
  object?: any
}

// The tool arguments are invalid
export type ToolResponseInvalid = {
  type: ToolResponseType.Invalid
  message: string
}

// Some error occurred when executing the tool
// e.g. network request error, IO error
export type ToolResponseError = {
  type: ToolResponseType.Error
  message: string
  // If true, the tool can be retried
  // e.g. network request error are generally retryable
  // but IO errors are not
  canRetry?: boolean
}

// The tool execution was interrupted
// e.g. user cancelled the tool execution
// or some security policy was violated
export type ToolResponseInterrupted = {
  type: ToolResponseType.Interrupted
  message: string
}

// Hand over the task to another agent
// e.g. hand over a coding task to the coder agent
export type ToolResponseHandOver = {
  type: ToolResponseType.HandOver
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

// Delegate the task to another agent
export type ToolResponseDelegate = {
  type: ToolResponseType.Delegate
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

export type ToolResponsePause = {
  type: ToolResponseType.Pause
  object: any
}

export type ToolResponse =
  | ToolResponseReply
  | ToolResponseExit
  | ToolResponseInvalid
  | ToolResponseError
  | ToolResponseInterrupted
  | ToolResponseHandOver
  | ToolResponseDelegate
  | ToolResponsePause

export type ToolHandler<T extends ToolInfo, P> = (
  provider: P,
  args: Partial<Record<T['parameters'][number]['name'], ToolParameterValue>>,
) => Promise<ToolResponse>
