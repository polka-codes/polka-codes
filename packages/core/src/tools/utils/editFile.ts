// Core logic for flexible file editing with before/after text anchors
// generated by polka.codes

/**
 * NOTE: Both `start_anchor` and `end_anchor` are **exclusive** markers.
 * They remain in the file; `new_text` is inserted *between* them.
 */
export interface EditOperation {
  start_anchor?: string
  end_anchor?: string
  new_text: string
  start_anchor_line_start?: number
  end_anchor_line_start?: number
}

export const START_OF_FILE = '<<<START_OF_FILE>>>'
export const END_OF_FILE = '<<<END_OF_FILE>>>'

/**
 * Apply multiple edit operations to file content using start/end text anchors (anchors are **exclusive** â€“ they stay in the file).
 * @param fileContent - The original file content
 * @param operations - Array of edit operations to apply
 * @returns The modified file content
 */
export const editFile = async (fileContent: string, operations: EditOperation[]): Promise<string> => {
  if (!operations || operations.length === 0) {
    throw new Error('At least one edit operation is required')
  }

  const originalLines = fileContent.split('\n')
  let updatedContent = fileContent

  // Process operations in order
  for (const operation of operations) {
    updatedContent = await applyEditOperation(updatedContent, operation, originalLines)
  }

  return updatedContent
}

async function applyEditOperation(fileContent: string, operation: EditOperation, originalLines: string[]): Promise<string> {
  const { start_anchor, end_anchor, new_text, start_anchor_line_start, end_anchor_line_start } = operation

  // Handle special cases for start/end of file
  if (start_anchor === START_OF_FILE && end_anchor === END_OF_FILE) {
    // Replace entire file
    return new_text
  }

  if (start_anchor === START_OF_FILE) {
    // Insert at start or replace from start to end_anchor
    if (!end_anchor) {
      return new_text + fileContent
    }
    const afterIndex = findTextWithHint(fileContent, end_anchor, end_anchor_line_start, originalLines)
    return new_text + fileContent.slice(afterIndex)
  }

  if (end_anchor === END_OF_FILE) {
    // Insert at end or replace from start_anchor to end
    if (!start_anchor) {
      return fileContent + new_text
    }
    const beforeIndex = findTextWithHint(fileContent, start_anchor, start_anchor_line_start, originalLines)
    const beforeEndIndex = beforeIndex + start_anchor.length
    return fileContent.slice(0, beforeEndIndex) + new_text
  }

  // Handle normal case with both start_anchor and end_anchor
  if (start_anchor && end_anchor) {
    const beforeIndex = findTextWithHint(fileContent, start_anchor, start_anchor_line_start, originalLines)
    const beforeEndIndex = beforeIndex + start_anchor.length
    const afterIndex = findTextWithHint(fileContent, end_anchor, end_anchor_line_start, originalLines, beforeEndIndex)

    return fileContent.slice(0, beforeEndIndex) + new_text + fileContent.slice(afterIndex)
  }

  // Handle case with only start_anchor (insert after)
  if (start_anchor) {
    const beforeIndex = findTextWithHint(fileContent, start_anchor, start_anchor_line_start, originalLines)
    const beforeEndIndex = beforeIndex + start_anchor.length
    return fileContent.slice(0, beforeEndIndex) + new_text + fileContent.slice(beforeEndIndex)
  }

  // Handle case with only end_anchor (insert before)
  if (end_anchor) {
    const afterIndex = findTextWithHint(fileContent, end_anchor, end_anchor_line_start, originalLines)
    return fileContent.slice(0, afterIndex) + new_text + fileContent.slice(afterIndex)
  }

  throw new Error('Either start_anchor or end_anchor must be specified')
}

function findTextWithHint(
  content: string,
  searchText: string,
  lineHint: number | undefined,
  originalLines: string[],
  startIndex = 0,
): number {
  // If we have a line hint, try searching around that area first
  if (lineHint && lineHint > 0 && lineHint <= originalLines.length) {
    const hintIndex = getLineStartIndex(originalLines, lineHint - 1) // Convert to 0-based
    const searchRadius = 5 // Search within 5 lines of the hint

    // Search in a window around the hint
    const windowStart = Math.max(0, hintIndex - searchRadius * 50) // Approximate chars per line
    const windowEnd = Math.min(content.length, hintIndex + searchRadius * 50)
    const windowContent = content.slice(windowStart, windowEnd)

    const relativeIndex = windowContent.indexOf(searchText)
    if (relativeIndex !== -1) {
      const absoluteIndex = windowStart + relativeIndex
      if (absoluteIndex >= startIndex) {
        return absoluteIndex
      }
    }
  }

  // Fallback to regular search from startIndex
  const index = content.indexOf(searchText, startIndex)
  if (index === -1) {
    throw new Error(`Could not find text: ${searchText}`)
  }
  return index
}

function getLineStartIndex(lines: string[], lineIndex: number): number {
  let index = 0
  for (let i = 0; i < lineIndex && i < lines.length; i++) {
    index += lines[i].length + 1 // +1 for newline character
  }
  return index
}
