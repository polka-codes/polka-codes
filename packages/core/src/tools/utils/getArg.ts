import type { ToolParameterValue } from '../../tool'

export const getString = <T extends string>(
  args: Partial<Record<T, string>> | ToolParameterValue,
  name: T,
  defaultValue?: string,
): string => {
  if (typeof args !== 'object' || Array.isArray(args)) {
    throw new Error(`Invalid argument type: ${name} ${args}`)
  }
  const ret = args[name] ?? defaultValue
  if (ret === undefined) {
    throw new Error(`Missing required argument: ${name}`)
  }
  if (typeof ret !== 'string') {
    throw new Error(`Invalid argument type: ${name} ${ret}`)
  }
  return ret
}

export const getStringArray = <T extends string>(
  args: Partial<Record<T, ToolParameterValue>>,
  name: T,
  defaultValue?: string[],
): string[] => {
  const ret = args[name]
  if (ret === undefined) {
    if (defaultValue === undefined) {
      throw new Error(`Missing required argument: ${name}`)
    }
    return defaultValue
  }
  if (ret === '') {
    return []
  }
  if (typeof ret !== 'string') {
    throw new Error(`Invalid argument type: ${name} ${ret}`)
  }
  return ret.split(',').map((s) => s.trim())
}

export const getBoolean = <T extends string>(args: Partial<Record<T, ToolParameterValue>>, name: T, defaultValue?: boolean): boolean => {
  const ret = args[name]
  if (ret === undefined) {
    if (defaultValue === undefined) {
      throw new Error(`Missing required argument: ${name}`)
    }
    return defaultValue
  }
  if (typeof ret !== 'string') {
    throw new Error(`Invalid argument type: ${name} ${ret}`)
  }
  switch (ret.toLowerCase()) {
    case 'true':
      return true
    case 'false':
      return false
    default:
      throw new Error(`Invalid argument value: ${name} ${ret}`)
  }
}

export const getInt = <T extends string>(args: Partial<Record<T, ToolParameterValue>>, name: T, defaultValue?: number): number => {
  const ret = args[name]
  if (ret === undefined) {
    if (defaultValue === undefined) {
      throw new Error(`Missing required argument: ${name}`)
    }
    return defaultValue
  }
  if (typeof ret !== 'string') {
    throw new Error(`Invalid argument type: ${name} ${ret}`)
  }
  const parsed = Number.parseInt(ret)
  if (Number.isNaN(parsed)) {
    throw new Error(`Invalid argument value: ${name} ${ret}`)
  }
  return parsed
}

export const getArray = <T extends string>(
  args: Partial<Record<T, ToolParameterValue>> | ToolParameterValue,
  name: T,
  defaultValue?: ToolParameterValue[],
): ToolParameterValue[] => {
  if (typeof args !== 'object' || Array.isArray(args)) {
    throw new Error(`Invalid argument type: ${name} ${args}`)
  }
  const ret = args[name]
  if (ret === undefined) {
    if (defaultValue === undefined) {
      throw new Error(`Missing required argument: ${name}`)
    }
    return defaultValue
  }
  if (Array.isArray(ret)) {
    return ret
  }
  return [ret]
}
