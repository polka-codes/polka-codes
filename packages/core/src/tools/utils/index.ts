export * from './editFile'
export * from './getArg'
export * from './replaceInFile'
