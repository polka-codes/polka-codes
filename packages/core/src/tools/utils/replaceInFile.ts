/* Example diff block format:
  <<<<<<< SEARCH
  // original text
  =======
  // replacement text
  >>>>>>> REPLACE
*/

export type ReplaceResult = {
  content: string
  status: 'no_diff_applied' | 'some_diff_applied' | 'all_diff_applied'
  appliedCount: number
  totalCount: number
}

export const replaceInFile = (fileContent: string, diff: string): ReplaceResult => {
  // Regex to match blocks of the form:
  // <<<<<<< SEARCH
  // (some lines)
  // =======
  // (some lines)
  // >>>>>>> REPLACE
  const blockPattern = /^\s*<<<<<+\s*SEARCH>?\s*\r?\n([\s\S]*?)\r?\n=======[ \t]*\r?\n([\s\S]*?)\r?\n?>>>>>+\s*REPLACE\s*$/gm

  // Parse diff blocks
  const blocks: { search: string; replace: string }[] = []
  for (let match = blockPattern.exec(diff); match !== null; match = blockPattern.exec(diff)) {
    blocks.push({ search: match[1], replace: match[2] })
  }
  if (blocks.length === 0) {
    throw new Error('No valid diff blocks found.')
  }

  // Helper: try to find the search text in fileContent with progressive relaxation
  const findAndReplace = (content: string, search: string, replace: string): string | null => {
    // 1) Direct exact match
    let index = content.indexOf(search)
    if (index !== -1) {
      return content.slice(0, index) + replace + content.slice(index + search.length)
    }

    // 2) Trim leading/ending whitespace in search and content
    const trimmedSearch = search.trim()
    const trimmedContent = content.trim()
    const offset = content.indexOf(trimmedContent) // to restore original indexing if found
    index = trimmedContent.indexOf(trimmedSearch)
    if (index !== -1) {
      // compute correct absolute index in the original content
      const absoluteIndex = offset + index
      return content.slice(0, absoluteIndex) + replace + content.slice(absoluteIndex + trimmedSearch.length)
    }

    // 3) Whitespace-agnostic match:
    //    Replace all consecutive whitespace in search and content with a single marker
    //    to see if there's a match ignoring whitespace diffs
    const normalizedSearch = trimmedSearch.replace(/\s+/g, ' ')
    const normalizedContent = trimmedContent.replace(/\s+/g, ' ')
    index = normalizedContent.indexOf(normalizedSearch)
    if (index !== -1) {
      // Find actual location in the original content. We do a rough approach here:
      // We know the substring's "normalized" start is at 'index' in normalizedContent.
      // A simple way is to walk through the original trimmedContent to find where that occurs.
      // For brevity, we'll do a naive re-scan and hope it's correct in typical cases.
      let runningIndex = 0
      let actualPos = offset
      for (const segment of trimmedSearch.replace(/\s+/g, ' ').split(' ')) {
        const segIndex = content.indexOf(segment, actualPos)
        if (segIndex === -1) {
          break // mismatch, won't happen if we truly found it
        }
        if (runningIndex === 0) {
          // First segment helps define the start
          actualPos = segIndex
        } else {
          // Move just after the segment to keep scanning
          actualPos = segIndex + segment.length
        }
        runningIndex++
      }

      // By the time weâ€™re done, actualPos should be the end of the matched substring.
      // We do a length calc for the final replacement index:
      // but we need the total length of trimmedSearch minus whitespace. We'll reconstruct:
      const strippedSearch = trimmedSearch.replace(/\s+/g, '')
      const endPos = actualPos // The end of the final segment
      const startPos = endPos - strippedSearch.length

      return content.slice(0, startPos) + replace + content.slice(endPos)
    }

    return null
  }

  let updatedFile = fileContent
  let appliedCount = 0
  const totalCount = blocks.length

  for (const { search, replace } of blocks) {
    const result = findAndReplace(updatedFile, search, replace)
    if (result !== null) {
      updatedFile = result
      appliedCount++
    }
  }

  let status: ReplaceResult['status']
  if (appliedCount === 0) {
    status = 'no_diff_applied'
  } else if (appliedCount < totalCount) {
    status = 'some_diff_applied'
  } else {
    status = 'all_diff_applied'
  }

  return {
    content: updatedFile,
    status,
    appliedCount,
    totalCount,
  }
}
