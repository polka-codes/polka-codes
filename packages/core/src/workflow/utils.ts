// generated by polka.codes
import { template } from 'lodash'
import type { Json, TemplatedString } from './types'

export const resolveTemplatedString = (
  templatedString: string | TemplatedString<{ [k: string]: Json }>,
  input: { [k: string]: Json },
): string => {
  if (typeof templatedString === 'string') {
    return templatedString
  }
  if (templatedString.type === 'function') {
    return templatedString.fn(input)
  }
  return template(templatedString.template)(input)
}
