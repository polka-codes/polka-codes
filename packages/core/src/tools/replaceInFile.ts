// generated by polka.codes
import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../tool'
import type { FilesystemProvider } from './provider'
import { replaceInFile } from './utils/replaceInFile'

export const toolInfo = {
  name: 'replace_in_file',
  description:
    'Request to replace sections of content in an existing file using SEARCH/REPLACE blocks that define exact changes to specific parts of the file. This tool should be used when you need to make targeted changes to specific parts of a file.',
  parameters: z
    .object({
      path: z.string().describe('The path of the file to modify').meta({ usageValue: 'File path here' }),
      diff: z
        .string()
        .describe(
          `One or more SEARCH/REPLACE blocks following this exact format:
\`\`\`
<<<<<<< SEARCH
[exact content to find]
=======
[new content to replace with]
>>>>>>> REPLACE
\`\`\`
Critical rules:
1. SEARCH content must match the associated file section to find EXACTLY:
    * Match character-for-character including whitespace, indentation, line endings
    * Include all comments, docstrings, etc.
2. SEARCH/REPLACE blocks will ONLY replace the first match occurrence.
    * Including multiple unique SEARCH/REPLACE blocks if you need to make multiple changes.
    * Include *just* enough lines in each SEARCH section to uniquely match each set of lines that need to change.
    * When using multiple SEARCH/REPLACE blocks, list them in the order they appear in the file.
3. Keep SEARCH/REPLACE blocks concise:
    * Break large SEARCH/REPLACE blocks into a series of smaller blocks that each change a small portion of the file.
    * Include just the changing lines, and a few surrounding lines if needed for uniqueness.
    * Do not include long runs of unchanging lines in SEARCH/REPLACE blocks.
    * Each line must be complete. Never truncate lines mid-way through as this can cause matching failures.
4. Special operations:
    * To move code: Use two SEARCH/REPLACE blocks (one to delete from original + one to insert at new location)
    * To delete code: Use empty REPLACE section`,
        )
        .meta({ usageValue: 'Search and replace blocks here' }),
    })
    .meta({
      examples: [
        {
          description: 'Request to replace sections of content in a file',
          input: {
            path: 'src/main.js',
            diff: `<<<<<<< SEARCH
import React from 'react';
=======
import React, { useState } from 'react';
>>>>>>> REPLACE

<<<<<<< SEARCH
function handleSubmit() {
  saveData();
  setLoading(false);
}

=======
>>>>>>> REPLACE

<<<<<<< SEARCH
return (
  <div>
=======
function handleSubmit() {
  saveData();
  setLoading(false);
}

return (
  <div>
>>>>>>> REPLACE`,
          },
        },
        {
          description: 'Request to perform a simple, single-line replacement',
          input: {
            path: 'src/config.js',
            diff: `<<<<<<< SEARCH
const API_URL = 'https://api.example.com';
=======
const API_URL = 'https://api.staging.example.com';
>>>>>>> REPLACE`,
          },
        },
        {
          description: 'Request to add a new function to a file',
          input: {
            path: 'src/utils.js',
            diff: `<<<<<<< SEARCH
function helperA() {
  // ...
}
=======
function helperA() {
  // ...
}

function newHelper() {
  // implementation
}
>>>>>>> REPLACE`,
          },
        },
        {
          description: 'Request to delete a block of code from a file',
          input: {
            path: 'src/app.js',
            diff: `<<<<<<< SEARCH
function oldFeature() {
  // This is no longer needed
}

=======
>>>>>>> REPLACE`,
          },
        },
      ],
    }),
  permissionLevel: PermissionLevel.Write,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, FilesystemProvider> = async (provider, args) => {
  if (!provider.readFile || !provider.writeFile) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to replace in file. Abort.',
    }
  }

  const parsed = toolInfo.parameters.safeParse(args)
  if (!parsed.success) {
    return {
      type: ToolResponseType.Invalid,
      message: `Invalid arguments for replace_in_file: ${parsed.error.message}`,
    }
  }
  const { path, diff } = parsed.data

  try {
    const fileContent = await provider.readFile(path, false)

    if (fileContent == null) {
      return {
        type: ToolResponseType.Error,
        message: `<replace_in_file_result path=\"${path}\" status=\"failed\" message=\"File not found\" />`,
      }
    }

    const result = replaceInFile(fileContent, diff)

    if (result.status === 'no_diff_applied') {
      return {
        type: ToolResponseType.Error,
        message: `<replace_in_file_result path=\"${path}\" status=\"failed\" message=\"Unable to apply changes\">
  <file_content path=\"${path}\">${fileContent}</file_content>
</replace_in_file_result>`,
      }
    }

    await provider.writeFile(path, result.content)

    if (result.status === 'some_diff_applied') {
      return {
        type: ToolResponseType.Reply,
        message: `<replace_in_file_result path=\"${path}\" status=\"some_diff_applied\" applied_count=\"${result.appliedCount}\" total_count=\"${result.totalCount}\">
  <file_content path=\"${path}\">${result.content}</file_content>
</replace_in_file_result>`,
      }
    }

    return {
      type: ToolResponseType.Reply,
      message: `<replace_in_file_result path=\"${path}\" status=\"all_diff_applied\" />`,
    }
  } catch (error) {
    return {
      type: ToolResponseType.Invalid,
      message: `Invalid arguments for replace_in_file: ${error}`,
    }
  }
}

export const isAvailable = (provider: FilesystemProvider): boolean => {
  return !!provider.readFile && !!provider.writeFile
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
