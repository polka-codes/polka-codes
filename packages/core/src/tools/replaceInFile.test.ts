// generated by polka.codes
import { describe, expect, type Mock, mock, spyOn, test } from 'bun:test'
import { type ToolHandler, ToolResponseType } from '../tool'
import type { FilesystemProvider } from './provider'
import { handler, isAvailable } from './replaceInFile'
import * as replaceInFileModule from './utils/replaceInFile'

describe('replaceInFile tool', () => {
  const mockProvider: FilesystemProvider = {
    readFile: mock((_path: string): Promise<string | undefined> => Promise.resolve('')),
    writeFile: mock((_path: string, _content: string): Promise<void> => Promise.resolve()),
  }

  test('isAvailable should return true when provider has readFile and writeFile', () => {
    expect(isAvailable(mockProvider)).toBe(true)
  })

  test('isAvailable should return false when provider is missing readFile', () => {
    const incompleteProvider = { writeFile: mock(() => Promise.resolve()) }
    expect(isAvailable(incompleteProvider)).toBe(false)
  })

  test('isAvailable should return false when provider is missing writeFile', () => {
    const incompleteProvider = { readFile: mock(() => Promise.resolve('')) }
    expect(isAvailable(incompleteProvider)).toBe(false)
  })

  test('handler should return error if provider is missing methods', async () => {
    const incompleteProvider = { writeFile: mock(() => Promise.resolve()) }
    const result = await (handler as ToolHandler<any, any>)(incompleteProvider, {})
    expect(result).toEqual({
      type: ToolResponseType.Error,
      message: {
        type: 'error-text',
        value: 'Not possible to replace in file.',
      },
    })
  })

  test('handler should return file not found error', async () => {
    ;(mockProvider.readFile as Mock<any>).mockResolvedValueOnce(undefined)
    const args = { path: 'nonexistent.txt', diff: '...' }
    const result = await (handler as ToolHandler<any, any>)(mockProvider, args)
    expect(result.type).toBe(ToolResponseType.Error)
    if (result.type === ToolResponseType.Error) {
      expect(result.message).toMatchSnapshot()
    }
  })

  test('handler should return no_diff_applied message', async () => {
    ;(mockProvider.readFile as Mock<any>).mockResolvedValueOnce('file content')
    spyOn(replaceInFileModule, 'replaceInFile').mockReturnValueOnce({
      status: 'no_diff_applied',
      appliedCount: 0,
      totalCount: 2,
      content: 'file content',
    })
    const args = { path: 'test.txt', diff: '...' }
    const result = await (handler as ToolHandler<any, any>)(mockProvider, args)
    expect(result).toMatchSnapshot()
  })

  test('handler should return some_diff_applied message', async () => {
    ;(mockProvider.readFile as Mock<any>).mockResolvedValueOnce('file content')
    spyOn(replaceInFileModule, 'replaceInFile').mockReturnValueOnce({
      status: 'some_diff_applied',
      appliedCount: 1,
      totalCount: 2,
      content: 'new content',
    })
    const args = { path: 'test.txt', diff: '...' }
    const result = await (handler as ToolHandler<any, any>)(mockProvider, args)
    expect(result.type).toBe(ToolResponseType.Reply)
    if (result.type === ToolResponseType.Reply) {
      expect(result.message).toMatchSnapshot()
    }
    expect(mockProvider.writeFile).toHaveBeenCalledWith('test.txt', 'new content')
  })

  test('handler should return all_diff_applied message', async () => {
    ;(mockProvider.readFile as Mock<any>).mockResolvedValueOnce('file content')
    spyOn(replaceInFileModule, 'replaceInFile').mockReturnValueOnce({
      status: 'all_diff_applied',
      appliedCount: 2,
      totalCount: 2,
      content: 'new content',
    })
    const args = { path: 'test.txt', diff: '...' }
    const result = await (handler as ToolHandler<any, any>)(mockProvider, args)
    expect(result.type).toBe(ToolResponseType.Reply)
    if (result.type === ToolResponseType.Reply) {
      expect(result.message).toMatchSnapshot()
    }
    expect(mockProvider.writeFile).toHaveBeenCalledWith('test.txt', 'new content')
  })
})
