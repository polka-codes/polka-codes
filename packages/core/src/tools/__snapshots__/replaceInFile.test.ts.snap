// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`replaceInFile tool handler should return file not found error 1`] = `"<replace_in_file_result path="nonexistent.txt" status="failed" message="File not found" />"`;

exports[`replaceInFile tool handler should return no_diff_applied message 1`] = `
{
  "message": 
"<replace_in_file_result path="test.txt" status="failed" message="Unable to apply changes">
  <file_content path="test.txt">file content</file_content>
</replace_in_file_result>"
,
  "type": "Error",
}
`;

exports[`replaceInFile tool handler should return some_diff_applied message 1`] = `
"<replace_in_file_result path="test.txt" status="some_diff_applied" applied_count="1" total_count="2">
  <file_content path="test.txt">new content</file_content>
</replace_in_file_result>"
`;

exports[`replaceInFile tool handler should return all_diff_applied message 1`] = `"<replace_in_file_result path="test.txt" status="all_diff_applied" />"`;
